File: openfeed-backend/openfeed/db/user_stories_hidden.py
from uuid import UUID

from openfeed.db.client import Client


def insert_hidden_story(db: Client, user_id: UUID, story_id: UUID) -> None:
    """Record that a user has hidden a story."""
    db.table("user_stories_hidden").insert(
        {"user_id": str(user_id), "story_id": str(story_id)}
    ).execute()


def get_hidden_story_ids(db: Client, user_id: UUID) -> set[UUID]:
    """Return the set of story_ids hidden by a user."""
    response = (
        db.table("user_stories_hidden")
        .select("story_id")
        .eq("user_id", str(user_id))
        .execute()
    )
    return {UUID(row["story_id"]) for row in response.data}

File: openfeed-frontend/lib/supabase/queries/user_stories_hidden.ts
import { SupabaseClient } from "@supabase/supabase-js";

import { Database } from "@/lib/supabase/supabase.types";

export async function getHiddenStoryIds(
  supabase: SupabaseClient<Database>,
  userId: string,
): Promise<Set<string>> {
  const { data, error } = await (supabase as any)
    .from("user_stories_hidden")
    .select("story_id")
    .eq("user_id", userId);

  if (error) throw new Error(error.message);
  return new Set((data as { story_id: string }[]).map((r) => r.story_id));
}

export async function hideStory(
  supabase: SupabaseClient<Database>,
  userId: string,
  storyId: string,
): Promise<void> {
  const { error } = await supabase
    .from("user_stories_hidden" as any)
    .insert({ user_id: userId, story_id: storyId });

  if (error) throw new Error(error.message);
}
